python/ray/util/multiprocessing/pool.py

```diff
@@ -186,46 +187,65 @@ def _add_object_ref(self, object_ref):
         self._object_refs.append(object_ref)
         self._results.append(None)
 
-    def add_object_ref(self, object_ref):
+    def add_object_ref(self, object_ref, final=False):
         self._new_object_refs.put(object_ref)
+        if final:
+            # To stop iterating, push an additional None object
+            self._new_object_refs.put(None)
 
     def run(self):
         unready = copy.copy(self._object_refs)
         aggregated_batch_results = []
-        while self._num_ready < self._total_object_refs:
+        winding_down = False
+        # Run for a specific number of objects if self._total_object_refs > 0
+        # Run until a 'final' (None) object has been pushed when iterating over an
+        # iterators (i.e. self._total_object_refs == 0)
+        while (self._total_object_refs == 0) or (
+            self._num_ready < self._total_object_refs
+        ):
             # Get as many new IDs from the queue as possible without blocking,
             # unless we have no IDs to wait on, in which case we block.
             while True:
                 try:
-                    block = len(unready) == 0
+                    # Block if we have no IDs to wait on *and* we are not winding down
+                    block = (len(unready) == 0) and not winding_down
                     new_object_ref = self._new_object_refs.get(block=block)
-                    self._add_object_ref(new_object_ref)
-                    unready.append(new_object_ref)
+                    # receiving a None object is the signal to stop
```

Comment:
`None` has become a kind of [magic value](https://en.wikipedia.org/wiki/Magic_number_(programming)). Can we add a class variable to `ResultThread` called `END_SENTINEL`, and set it equal to `None`. Whenever, we want to add a `None` or check for `None` that represents the end of `_new_object_refs`, we can use `ResultThread.END_SENTINEL` instead.

It would also be nice to add a comment explaining `END_SENTINEL`'s behavior:

```python
# Signals that there are no more new object_refs.
# When this thread gets the END_SENTINEL from
# self._new_object_refs, it winds down.
END_SENTINEL =  None
```

Reply:
Great suggestion. I have added END_SENTINEL as a class variable, with the above explanation.